### ISSUE TITLE
Master: Delete in JSON and RDF seems not working as it should.

### ISSUE THREAD

**Issue by MichelDiz** [dgraph-io/dgraph#3477]
If you suspect this could be a bug, follow the template.

- What version of Dgraph are you using?
Master (works on 1.0.14)

- Steps to reproduce the issue (command/config used to run Dgraph).

This was discovered in part by doing tests to https://github.com/dgraph-io/dgraph/issues/3447 
and also I've noticed this few days ago, but I thought It was some issue with my machine. Now @mangalaman93  Also is going through this. So by replication/reproduction this can be suspect.

## Schema

```
name: string @index(term, exact) @lang .
age: int @index(int) .
friend: [uid] @count .
parent_to: [uid] @count .
role: string @lang .

```
## Mutation

```bash
curl -H "Content-Type: application/json" localhost:8080/mutate?commitNow=true -X POST -d $'
{
  "set": [
 add here this JSON :
https://github.com/dgraph-io/dgraph/blob/master/dgraph/cmd/live/load-json/family.json
    ]
}
' | jq
```

## Query

```bash
curl -H "Content-Type: application/graphqlpm" localhost:8080/query -X POST -d $'
{
  q(func: eq(name, "Bart")){
    uid
    name
  }
}
' | jq
```

## Delete

```bash
curl -H "Content-Type: application/json" localhost:8080/mutate?commitNow=true -X POST -d $'
{
  "delete": [
    {
      "uid": "0x5" #Delete Bart
    }
  ]
}
' | jq
```

```
curl -H "Content-Type: application/rdf" localhost:8080/mutate?commitNow=true -X POST -d $'
{
  delete {
      <0x5> * * .
  }
}
' | jq
```

After try to delete Bart. You gonna see that the operation throw "success" but Bart still there.

- Expected behaviour and actual result.

**Comment by mangalaman93** [comment]
I am able to delete a node in release binary of v1.0.14. Let me run git-bisect (or equivalent) to figure out the commit that introduced this issue.

**Comment by mangalaman93** [comment]
Result of git bisect. (current master: 14fa8274499b01602343a90a726a7744db691fd5)

```
d7f7d5ffa99d7b50dd43974d8f9447ffabf736d6 is the first bad commit
commit d7f7d5ffa99d7b50dd43974d8f9447ffabf736d6
Author: Martin Martinez Rivera <martinmr@users.noreply.github.com>
Date:   Thu May 9 12:15:35 2019 -0700

    Remove _predicate_ from Dgraph. (#3262)
    
    From now on, expand queries will rely solely on the type of the node.
    This change removes all mentions of the _predicate_ internal pred that
    was previously being used to serve this type of queries.
    
    There is also another small change to allow the value of a list variable
    to be used as a list of predicates in expand queries.
    
    Most of the changes due to removing or replacing tests. These changes
    are due to one of the following reasons.
    
    1. Tests that were checking for the state of _predicate_ and are no
    longer valid.
    2. Tests that were doing checks that should be done inside of the query
    package. They have been removed if there was an existing query that was
    doing the same or a new test has been added if no similar test existed.
    3. Some of the expand tests inside the query package were removed
    because there are tests verifying the expand functionality using types.
    Since types are now the only way to make expand queries work, these
    tests became obsolete.
    
    Verified jepsen tests work and that the 21 million dataset works as expected
    by manually adding types to certain nodes. Types and type definitons
    will be added to the entire dataset later.

:040000 040000 06336d3697fb900cc94cf6a0f5c67ec09e786418 ad1bf1cef75aebc95a95d219fb881b59087cbe07 M	dgraph
:040000 040000 354a158e95d55bce2aafc69a3dfef1e8eb5e2ad3 d6593c8666967a5cd64cd85166a7e39d2845a9fe M	gql
:040000 040000 08421e199654ef98d3db246754be70a33ed06174 64715db38ca69b55308b692884de29b19d8fa552 M	posting
:040000 040000 098a4f635a157ca7e1adde380e2d4f2592828e9d eb00f2b05865f28cd1421737f6c7ee107e66506c M	query
:040000 040000 281970aeea7ac40b454ef564fdb74960dd11d62e e7846b80a4a2816ee02329c56c43cf66d18d2ca9 M	schema
:040000 040000 cd649d7eb9a527c939494843861c1d8021c45be5 3032a7de12f5c6742c1905a2dea1a34d2ed9d621 M	systest
:040000 040000 8d282f43aa4be42ae54d3c8954c06bf688dde0a3 98344b207bb28a7cbaff0e6fe23e9892952faf22 M	worker
:040000 040000 c4b7e2ef2f74eb5e3f65ab4883194d6a5694c4ae 7026233296d30dca75a8d553905cf299003b708d M	x
:040000 040000 60c3a02466c10088b9c0dae7bb4454a6889746bd 94a14494bb7ab84eafedfce2149f13727c72107e M	z
``

**Comment by mangalaman93** [comment]
bisect logs -
```
aman@dgraph:~/gocode/src/github.com/dgraph-io/dgraph$ git bisect log
git bisect start
# bad: [14fa8274499b01602343a90a726a7744db691fd5] Add basic test utils for creating minio clients. (#3459)
git bisect bad 14fa8274499b01602343a90a726a7744db691fd5
# good: [913c9208191bfb94d661117852703f8e3c6551f7] Add timeout option while running queries
git bisect good 913c9208191bfb94d661117852703f8e3c6551f7
# good: [62580a1b151511f090f86b4a8885303e8f2e5402] Add support for JSON export (#3309)
git bisect good 62580a1b151511f090f86b4a8885303e8f2e5402
# bad: [b36f402ad1986016c9eff8fdb41ea82f37737b47] Fix race condition in numShutDownSig in alpha
git bisect bad b36f402ad1986016c9eff8fdb41ea82f37737b47
# good: [9086d00252e0fed3460f15c30f7249d69ad40729] Revert "Fix issue caused by the new RaftKey (#3370)"
git bisect good 9086d00252e0fed3460f15c30f7249d69ad40729
# good: [bfcf784dc7d252bff96250edb96685a50785164c] Bug Fix: Avoid Dgraph cluster getting stuck in infinite leader election (#3391)
git bisect good bfcf784dc7d252bff96250edb96685a50785164c
# bad: [390d57550c22e1f6d9f2809cb85791f5417b9f1c] Adding the migration tool (#3295)
git bisect bad 390d57550c22e1f6d9f2809cb85791f5417b9f1c
# bad: [d7f7d5ffa99d7b50dd43974d8f9447ffabf736d6] Remove _predicate_ from Dgraph. (#3262)
git bisect bad d7f7d5ffa99d7b50dd43974d8f9447ffabf736d6
# good: [cd1806359c79759c2ad02d0be025066d0ec81a1f] Add DockerCp to docker util functions. (#3390)
git bisect good cd1806359c79759c2ad02d0be025066d0ec81a1f
# first bad commit: [d7f7d5ffa99d7b50dd43974d8f9447ffabf736d6] Remove _predicate_ from Dgraph. (#3262)
```

**Comment by mangalaman93** [comment]
I have narrowed down the issue to the function `expandEdges` in file `query/mutation.go`. The function previously was expanding the `*` to all possible predicates. In current scenario, that doesn't seem to happen any more. I thought the function `GetAllPredicates` in `query/query.go` should help us find all predicates. That didn't help either. @martinmr ideas?

**Comment by mangalaman93** [comment]
Seems like this is expected behaviour. I will explain in my next comment.

**Comment by mangalaman93** [comment]
@manishrjain I still do not feel comfortable with the idea that on `delete { <uid> * * }` we do not delete all the outgoing predicates connected to `<uid>`. Instead of putting this disclaimer that we only delete typed predicated in this scenario, I would prefer a disclaimer that this operation could be slow because we are trying to delete all outgoing predicates. Assigning to you to make a decision.

**Comment by MichelDiz** [comment]
I think this is about "incoming edges" right? When you delete a node you also delete outgoing edges as far I know. Cuz the outgoing edges are in the target (from the operation) node. So incoming edges are in other nodes.

I think we could have a possible way to do this if we have "@reverse" in incoming edges. So this would be kind of easy to delete them. Without "@reverse" it's hard, you would use a lot of resources from the DB just to find some dangling edges. If this impacts performance, users would complain about it.

**Comment by mangalaman93** [comment]
@MichelDiz we are not deleting the outgoing edges (predicates) except the ones with the type information. Precisely, this issue.

**Comment by MichelDiz** [comment]
Humm, so this is related to the new type system?

**Comment by mangalaman93** [comment]
We have removed the `_predicate_` predicate now which is what used to keep the list of outgoing edges (predicates). Now, that we do not have that information, the only approach I could think of is to try all possible predicates. We do keep the type information on each node, so, we do delete the predicates in related to that type. I hope this clarifies what is happening.

**Comment by martinmr** [comment]
Deletion is working as intended. Starting from 1.1 `_predicate_` has been deleted so the only way to implement this operation is to expand the predicates based on the types of the node. I have added documentation for the type system but I might have missed updating the documentation on this feature. I'll submit a PR to fix the docs.

**Comment by manishrjain** [comment]
I've already explained my reasoning in the call we had. Not sure what more I could add.

Dgraph is a predicate based system. In our storage layer, we don't have the concept of a node, per se. Scanning over the entire database, which could be terabytes of data, to look for any edge which has that node isn't scalable. Instead, the user can easily determine what edges could be coming into a node, based on its type. And @martinmr 's change already looks at the type of the node to determine the outgoing edges, to deduce which combinations to try for deletion.

In fact, in the new type system (similar to GraphQL), a type is generally expected to have at least one "required" field. And any results of that type who don't have that field, would automatically be removed from the list of results. So, a uid which has been deleted wouldn't be returned anyway, even if there are pointers to it within the DB.

TLDR: Removal of these dangling pointers with DB-wide scanning isn't something Dgraph should do. It is up to the user to send Dgraph right instructions to either remove those or eliminate those from the result set by setting their schema correctly.

**Comment by martinmr** [comment]
Link to docs PR: https://github.com/dgraph-io/dgraph/pull/3489

**Comment by MichelDiz** [comment]
@martinmr  I've get confused about the PR #3489, it's not about this issue. The issue here on 3477 is that we can't actually execute a `S * *` operation on master. Is perfect possible to run `S P *`, this last op works great.  This part about "outgoing" edges was just a side comment.

After seeing your answer I cleaned all Dgraph files and did again "go get dgraph (...)". And then I did another build and cleaned up all my test files. And yet the problem persisted.

At first I believe that removing "`_predicate_`" also removed the possibility of performing `S * *` operations. Cuz `S P *` is possible tho, and also because I saw this behavior happen with ExpandAll - So it's related to "`_predicate_`". Or maybe you have another way to execute the `S * *` operation not yet introduced in the documentation. A `S <type> *`??

**Comment by MichelDiz** [comment]
The issue about "incoming edges / dangling edges" is https://github.com/dgraph-io/dgraph/issues/3447

**Comment by martinmr** [comment]
My PR is about how to get `S * *` working.

**Comment by MichelDiz** [comment]
Well this below is right? didn't worked.
It didn't deleted Lisa.

Please point me out a test file to this type of operation, so I can dig more.

### Type
```
type familyMember {
    name: string
    age: int
    role: string
    parent_to: uid
    aka: string
    carries: string
    sibling_of: uid
}
```

### Query
```
{
  q(func: eq(name, "Lisa")){
    uid
    name
    age
    role
    parent_to {
      uid
    }
    aka
    carries
    sibling_of {
      uid
    }
  }
}
```
### Delete OP in JSON
```
curl -H "Content-Type: application/json" localhost:8080/mutate?commitNow=true -X POST -d $'
{
  "delete": [
    {
      "uid": "0x2"
    }
  ]
}
' | jq
```
### Delete OP in JSON

```
curl -H "Content-Type: application/rdf" localhost:8080/mutate?commitNow=true -X POST -d $'
{
  delete {
      <0x2> * * .
  }
}
' | jq
```

**Comment by MichelDiz** [comment]
HUmmmm! I did not pay much attention to the text because I thought it was about clarifications about "dangling edges"...

So I have to first set the type of the node manually - that was I missing all the time. 

BTW: So this needs to go to all file tests we have (wow, a lot of work incoming).

```
{
  set {
      <0x2> <dgraph.type> "familyMember" .
  }
}
```

So, after defining the typo of my Node, not only ExpandAll worked as `S * *` too.

![image](https://user-images.githubusercontent.com/6570955/58671618-a2c8cb00-8319-11e9-8412-bc102ee280c3.png)
subtitles: Now I get it! Now I got it!
Now all the pieces fit together!
